django/contrib/admin/views/autocomplete.py

```diff
@@ -0,0 +1,169 @@
+import sys
+
+from django.core.exceptions import PermissionDenied
+from django.core.paginator import EmptyPage, Page, PageNotAnInteger, Paginator
+from django.db import router
+from django.http import Http404, JsonResponse
+from django.utils.encoding import force_text
+from django.utils.functional import cached_property
+from django.views.generic.list import BaseListView
+
+
+class AutocompletePage(Page):
+    def has_next(self):
+        """
+        Return true until queryset is smaller than the requested page size.
+
+        This avoids calling calculating the whole page count.
+        """
+        return self.object_list.count() >= self.paginator.per_page
+
+
+class AutocompletePaginator(Paginator):
+    """
+    A Django Paginator for big data sets.
+
+    Because users mostly use the first pages, it doesn't matter
+    if the page count is correct or not, as long as it's high enough.
+
+    Count() queries can take time, because they want to be exact.
+
```

Comment:
Remove newlines here, paragraphs are to separate thoughts, those belong logically together